client/components/tinymce/plugins/mentions/suggestion-list.jsx

```diff
@@ -0,0 +1,91 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+/**
+ * Internal dependencies
+ */
+import PopoverMenu from 'components/popover/menu';
+import PopoverMenuItem from 'components/popover/menu-item';
+import Suggestion from './suggestion';
+
+const SuggestionList = React.createClass( {
+	displayName: 'SuggestionList',
+
+	propTypes: {
+		query: React.PropTypes.string,
+		suggestions: React.PropTypes.array,
+		isVisible: React.PropTypes.bool,
+		onClick: React.PropTypes.func,
+		onClose: React.PropTypes.func
+	},
+
+	getDefaultProps() {
+		return {
+			query: '',
+			suggestions: [],
+			isVisible: false
+		};
+	},
+
+	getMatchingSuggestions() {
+		const { suggestions, query } = this.props;
+		const matcher = new RegExp( '^' + query + '| ' + query, 'ig' ); // Start of string or preceded by a space.
+		let matchingSuggestions = suggestions;
+
+		if ( query.length > 0 ) {
+			matchingSuggestions = [];
+
+			for ( let i = 0, len = suggestions.length; i < len; i++ ) {
+				const suggestion = suggestions[ i ];
+				const name = suggestion.name || suggestion.user_login + ' ' + suggestion.display_name;
+
+				if ( name.toLowerCase().match( matcher ) ) {
+					matchingSuggestions.push( suggestion );
+				}
+			}
+		}
+
+		return matchingSuggestions.slice( 0, 10 );
+	},
+
+	render() {
+		if ( ! this.props.isVisible || ! this.props.suggestions ) {
+			return null;
+		}
+
+		const suggestions = this.getMatchingSuggestions();
+
+		if ( suggestions.length > 0 ) {
+			return (
+				<div ref="mention">
+					<PopoverMenu
+						className="suggestions"
+						context={ this.refs && this.refs.mention }
+						isVisible={ this.props.isVisible }
+						onClose={ this.props.onClose }>
+							{ suggestions.map( ( suggestion ) => {
+								return (
+									<PopoverMenuItem
+										key={ 'user-suggestion-' + suggestion.ID }
+										onClick={ this.props.onClick.bind( null, suggestion ) }>
```

Comment:
Sorry I came off too much on the offensive. You're right that it's made more interesting by the `map` and the need to pass suggestion-specific data. I played around with it a bit, and I think you could achieve the effect without too much ugliness by having an instance-bound click handler as the `onClick` prop and a `data-suggestion-id` (or index) attribute which the click handler uses to look up the `suggestion` object. Not quite as direct as `bind`, but there's a small gain to be had.

This should illustrate it: http://jsbin.com/kalaronoju/edit?js,console,output
(Note too the renders in console, since new function in render map would ruin gains from PureComponent)

I like the rule, but it does surface many interesting edge cases. Might be good to discuss whether it's worth keeping with that in mind, and if so, where we can document how to handle said cases.